### PR TITLE
Add py-pennylane-lightning-kokkos package.

### PR DESCRIPTION
**Pull request by vincentmr**

### CODE REVIEW

File: var/spack/repos/builtin/packages/py-pennylane-lightning/package.py
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-pennylane-lightning/package.py]: The most recent stable release is already preferred by default. You can't prefer multiple versions.

File: var/spack/repos/builtin/packages/py-pennylane-lightning/package.py
Comment (mlxd) [inline var/spack/repos/builtin/packages/py-pennylane-lightning/package.py]: Do we need to have changes in the Py-PennyLane-Lightning file for this? I think the new package should be pretty orthogonal to this one

File: var/spack/repos/builtin/packages/py-pennylane/package.py
Comment (mlxd) [inline var/spack/repos/builtin/packages/py-pennylane/package.py:13]: Same here as with Py-PennyLane-Lightning. Do we need to change this file?

**Review by mlxd — COMMENTED** [review]
Thanks @vincentmr though quick questions:

File: var/spack/repos/builtin/packages/py-pennylane-lightning/package.py
Comment (vincentmr) [reply var/spack/repos/builtin/packages/py-pennylane-lightning/package.py]: Good, I'll remove that. For some reason, my `spack` would select `master` even after uninstalling everything and reconcretizing.

File: var/spack/repos/builtin/packages/py-pennylane/package.py
Comment (vincentmr) [reply var/spack/repos/builtin/packages/py-pennylane/package.py:13]: No strictly necessary. I thought I would just get all `py-pennylane` packages in-line as possible.

File: var/spack/repos/builtin/packages/py-pennylane-lightning/package.py
Comment (vincentmr) [reply var/spack/repos/builtin/packages/py-pennylane-lightning/package.py]: Sorry about this. I merged `gitlab/develop` first and I think I got lost in some conflicts.

**Comment by vincentmr** [comment]
@adamjstewart Should we still follow the guidelines from [this page](https://github.com/spack/spack-infrastructure/blob/main/docs/deferred_pipelines.md). If so, I was wondering whether I'm supposed to find a tag `latest-tested-develop` in the Gitlab repo.

**Comment by adamjstewart** [comment]
You don't need to do anything. All that page is saying is that the GitLab pipelines for this PR won't run until the pipeline for the commit from develop that this PR branched off has finished.

File: var/spack/repos/builtin/packages/py-pennylane-lightning-kokkos/package.py
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-pennylane-lightning-kokkos/package.py]:
```suggestion
    depends_on("py-pennylane@0.28:", type=("build", "run"))
```

File: var/spack/repos/builtin/packages/py-pennylane-lightning-kokkos/package.py
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-pennylane-lightning-kokkos/package.py]: I'm surprised to see ninja and pip are required to run the package, is that correct?

File: var/spack/repos/builtin/packages/py-pennylane-lightning-kokkos/package.py
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-pennylane-lightning-kokkos/package.py]: According to CMakeLists.txt this should be:
```suggestion
    depends_on("cmake@3.20:", type="build")
```
but if you know better than that file that's fine too.

File: var/spack/repos/builtin/packages/py-pennylane-lightning-kokkos/package.py
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-pennylane-lightning-kokkos/package.py]: Should prob be build/run to match other packages

**Review by adamjstewart — COMMENTED** [review]
I'm making the same assumptions as I made on your last PR, that you know your own software better than me. So my reviews are just suggestions of things that looked fishy, feel free to ignore me if I'm wrong.

File: var/spack/repos/builtin/packages/py-pennylane-lightning-kokkos/package.py
Comment (vincentmr) [reply var/spack/repos/builtin/packages/py-pennylane-lightning-kokkos/package.py]: `0.28` is the lowest available version so I didn't specify it, but you're right it will make the build more robust.

I don't need `py-pennylane` during the build phase, but I need it's binary `pl-device-test` to perform some post-installation tests. So do you think `type=("run", "test")` would be correct here?

File: var/spack/repos/builtin/packages/py-pennylane-lightning-kokkos/package.py
Comment (vincentmr) [reply var/spack/repos/builtin/packages/py-pennylane-lightning-kokkos/package.py]: I think you're right.

File: var/spack/repos/builtin/packages/py-pennylane-lightning-kokkos/package.py
Comment (adamjstewart) [reply var/spack/repos/builtin/packages/py-pennylane-lightning-kokkos/package.py]: Technically `type=("run", "test")` is more correct, although we usually use `type=("build", "run")` for all Python packages needed at run-time. "build" or "test" is needed so the import tests pass. I'll let you decide which you want to use here. I'm hoping that someday "run" will imply "test" for post-install tests. Maybe @tldahlgren has thoughts on that.

Comment (vincentmr) [reply var/spack/repos/builtin/packages/py-pennylane-lightning-kokkos/package.py]: I see, after reading the doc section on dependencies, your suggestion makes sense. I'll change it.

File: var/spack/repos/builtin/packages/py-pennylane-lightning-kokkos/package.py
Comment (vincentmr) [reply var/spack/repos/builtin/packages/py-pennylane-lightning-kokkos/package.py]: You're correct, this is an obsolete requirement. I'll remove it.

File: var/spack/repos/builtin/packages/py-pennylane-lightning-kokkos/package.py
Comment (vincentmr) [reply var/spack/repos/builtin/packages/py-pennylane-lightning-kokkos/package.py]: Done.

**Review by adamjstewart — DISMISSED** [review]
LGTM. If it looks fine to the rest of the team let me know and I'll merge.

**Comment by vincentmr** [comment]
> LGTM. If it looks fine to the rest of the team let me know and I'll merge.

Thanks for the swift review.

**Review by marcodelapierre — COMMENTED** [review]
Good to me, as far as comments from the other reviewers are addressed.

File: var/spack/repos/builtin/packages/py-pennylane/package.py
Comment (AmintorDusko) [inline var/spack/repos/builtin/packages/py-pennylane/package.py]: Because we have the 0.29.1 bug-fix release, we prefer not to add version 0.29.0, since it is already deprecated.

File: var/spack/repos/builtin/packages/py-pennylane/package.py
Comment (AmintorDusko) [inline var/spack/repos/builtin/packages/py-pennylane/package.py:20]: Hey @vincentmr, I don't get why we have a change in the hashes. Do you happen to know?

Comment (vincentmr) [reply var/spack/repos/builtin/packages/py-pennylane/package.py:20]: Yes, that's because I changed the source from `pypi` to `github`, as in the other py-penny packages. If you prefer I can revert everything in this file.

File: var/spack/repos/builtin/packages/py-pennylane/package.py
Comment (vincentmr) [reply var/spack/repos/builtin/packages/py-pennylane/package.py]: Do you prefer removing such versions or marking them `deprecated=True`?

File: var/spack/repos/builtin/packages/py-pennylane/package.py
Comment (AmintorDusko) [reply var/spack/repos/builtin/packages/py-pennylane/package.py:20]: I will defer to @mlxd  to decide that. I'm good with both, but I only tested the previous way.

File: var/spack/repos/builtin/packages/py-pennylane/package.py
Comment (AmintorDusko) [reply var/spack/repos/builtin/packages/py-pennylane/package.py]: In this case, the best course of action is to just remove it.

File: var/spack/repos/builtin/packages/py-pennylane/package.py
Comment (mlxd) [reply var/spack/repos/builtin/packages/py-pennylane/package.py:20]: Since Github will always be our source of truth and where we create releases, let's favour that across the packages until a time where we can move all to PyPI.

Comment (adamjstewart) [reply var/spack/repos/builtin/packages/py-pennylane/package.py:20]: For most Python packages we prefer PyPI (https://spack.readthedocs.io/en/latest/build_systems/pythonpackage.html#pypi-vs-github), but if you need files that aren't included in the PyPI tarball then it's fine to use GitHub.

Comment (mlxd) [reply var/spack/repos/builtin/packages/py-pennylane/package.py:20]: The long-plan is to improve the source distributions on PyPI so we can build with the same confidence as the Github releases. Happy to update to PyPI tags once this work is finished across all the packages.

**Comment by mlxd** [comment]
Thanks for the work on this @vincentmr 
No blockers from me. We can move all packages to PyPI once we update the source-only distributions for the packages in a follow up release around the 0.30/0.31 release timeline.

**Review by mlxd — DISMISSED** [review]

File: var/spack/repos/builtin/packages/py-pennylane-lightning-kokkos/package.py
Comment (AmintorDusko) [inline var/spack/repos/builtin/packages/py-pennylane-lightning-kokkos/package.py]: @vincentmr and @mlxd, could you please double-check if we'll proceed with this maintainers list?

Comment (adamjstewart) [reply var/spack/repos/builtin/packages/py-pennylane-lightning-kokkos/package.py]: We can add all of you if you all want to be listed. This is just a list of people who will be automatically asked to review PRs for this package.

Comment (AmintorDusko) [reply var/spack/repos/builtin/packages/py-pennylane-lightning-kokkos/package.py]: Sure! Thank you, @adamjstewart.
@vincentmr, feel free to add my name to the list.

Comment (mlxd) [reply var/spack/repos/builtin/packages/py-pennylane-lightning-kokkos/package.py]: I'm fine to have @AmintorDusko  and @vincentmr be the maintainers of this package.

**Review by AmintorDusko — APPROVED** [review]
Thank you, @vincentmr for your excellent job!

**Review by adamjstewart — APPROVED** [review]
From what I can tell it looks like all maintainers have approved, time to merge!

**Comment by AmintorDusko** [comment]
Thank you, @adamjstewart for your quick review!

**Pull request merged**
